src/libraries/ArrayUtil.sol

```diff
@@ -0,0 +1,565 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {SafeCast} from "openzeppelin/utils/math/SafeCast.sol";
+
+library ArrayUtil {
+    using SafeCast for uint256;
+    using SafeCast for int256;
+
+    error IndexOutOfBounds();
+
+    /**
+     * @dev Returns minimal element in array
+     * @return m
+     */
+    function min(int256[] memory x) internal pure returns (int256 m) {
+        m = x[0];
+        for (uint256 i; i < x.length; i++) {
+            if (x[i] < m) {
+                m = x[i];
+            }
+        }
+    }
+
+    function min(uint256[] memory x) internal pure returns (uint256) {
+        return min(toInt256(x)).toUint256();
+    }
+
+    // /**
+    //  * @dev Returns minimal element's index
+    //  * @return m
+    //  */
+    // function imin(uint256[] memory x) internal pure returns (uint256 m) {
+    //     m = 0;
+    //     for (uint256 i; i < x.length; i++) {
+    //         if (x[i] < x[m]) {
+    //             m = i;
+    //         }
+    //     }
+    //     return m;
+    // }
+
+    /**
+     * @dev Returns maximal element in array
+     * @return m
+     */
+    function max(int256[] memory x) internal pure returns (int256 m) {
+        m = x[0];
+        for (uint256 i; i < x.length; i++) {
+            if (x[i] > m) {
+                m = x[i];
+            }
+        }
+    }
+
+    function max(uint256[] memory x) internal pure returns (uint256) {
+        return max(toInt256(x)).toUint256();
+    }
+
+    /**
+     * @dev Returns maximal elements comparedTo value
+     * @return y array
+     */
+    function maximum(int256[] memory x, int256 z) internal pure returns (int256[] memory y) {
+        y = new int256[](x.length);
+        for (uint256 i; i < x.length; i++) {
+            if (x[i] > z) y[i] = x[i];
+            else y[i] = z;
+        }
+    }
+
+    // /**
+    //  * @dev Returns maximal element's index
+    //  * @return m maximal
+    //  */
+    // function imax(uint256[] memory x) internal pure returns (uint256 m) {
+    //     for (uint256 i; i < x.length; i++) {
+    //         if (x[i] > x[m]) {
+    //             m = i;
+    //         }
+    //     }
+    // }
+
+    /**
+     * @dev Removes element at index
+     * @return y new array
```

Comment:
We should add clear comment about whether this function change the memory in place or not